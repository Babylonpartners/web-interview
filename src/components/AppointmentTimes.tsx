import * as React from 'react'
import { headerStyles } from '../styles/styles';
import { Responsive } from 'semantic-ui-react';
import { AppointmentTimeMobButton } from './AppointmentTimeMobButton';
import { AppointmentTimeButton } from './AppointmentTimeButton';

interface Props {
    handleChange: any
    selectedConsultantType: any
    selectedAppointmentTime: any
    availableSlots: any
}

export class AppointmentTimes extends React.PureComponent <Props> {

    render() {

            // calculate matching slots

            let slots = []
            let slotData = this.props.availableSlots
                
            for (let i = 0, len=slotData.length ; i < len; i++) {
                let conType = slotData[i]['consultantType' as any]
                for ( let j = 0; j < conType.length; j++) {
                if (
                    slotData[i]['consultantType' as any][j] ===
                    this.props.selectedConsultantType
                ) {
                    slots.push(this.props.availableSlots[i])
                }
                }
            } 


        const monthNames = ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
        "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
        ];

        if(slots.length < 1) return <p data-testid="appointmentTimesLoading">Loading...</p>
        return (
            <div data-testid="appointmentTimesComponent">
            <Responsive maxWidth={750}>
            <strong
                style={headerStyles}
            >
                    Date and time
            </strong>
            <br />
            <br />
              <>
              {slots.map((slot: any) => (
                <React.Fragment
                    key={slot.id}
                >
                  <AppointmentTimeMobButton
                    handleChange={this.props.handleChange}
                    selectedAppointmentTime={this.props.selectedAppointmentTime}
                    slot={slot}
                    monthNames={monthNames}
                  />
                </React.Fragment>
               ))}
                </>
                </Responsive>


                <Responsive minWidth={751}>
                    <strong
                        style={headerStyles}
                    >
                            Date and time
                    </strong>
                    <br />
                    <br />
                    
                    <>
                    {slots.map((slot: any) => (
                        <React.Fragment
                            key={slot.id}
                        >
                        <AppointmentTimeButton 
                            handleChange={this.props.handleChange}
                            selectedAppointmentTime={this.props.selectedAppointmentTime}
                            slot={slot}
                            monthNames={monthNames} 
                        />
                        </React.Fragment>
                    ))}
                        </>
                </Responsive>
          </div>
        );
    }
}