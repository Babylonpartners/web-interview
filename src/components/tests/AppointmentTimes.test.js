import React from 'react'
import { render, cleanup } from '@testing-library/react'
import { AppointmentTimes } from '../AppointmentTimes'

afterEach(() => {
  cleanup()
  console.error.mockClear()
})

console.error = jest.fn()

const slots = [
  {
    id: 1,
    consultantType: ['gp'],
    appointmentType: ['audio', 'video'],
    time: '2019-11-27T10:11:00.000Z',
  },
  {
    id: 2,
    consultantType: ['specialist', 'gp'],
    appointmentType: ['audio', 'video'],
    time: '2019-12-01T14:16:30.000Z',
  },
  {
    id: 2,
    consultantType: ['specialist', 'gp'],
    appointmentType: ['audio', 'video'],
    time: '2019-12-01T14:16:30.000Z',
  },
]

test('<AppointmentTimes /> consultant type NOT selected', () => {
  const { getByTestId, queryByTestId, container } = render(
    <AppointmentTimes
      handleChange
      selectedConsultantType
      selectedAppointmentTime
      availableSlots={slots}
    />
  )
  expect(container.firstChild).toMatchSnapshot()
  expect(getByTestId('appointmentTimesLoading')).toBeTruthy()
  expect(queryByTestId('appointmentTimesComponent')).toBeFalsy()
})

test('<AppointmentTimes /> consultant type selected', () => {
  const { getByTestId, queryByTestId } = render(
    <AppointmentTimes
      handleChange
      selectedConsultantType={'gp'}
      selectedAppointmentTime
      availableSlots={slots}
    />
  )

  expect(queryByTestId('appointmentTimesLoading')).toBeFalsy()
  expect(getByTestId('appointmentTimesComponent')).toBeTruthy()
})

test('<AppointmentTimes /> number items unselected', () => {
  const { getAllByTestId } = render(
    <AppointmentTimes
      handleChange
      selectedConsultantType={'gp'}
      selectedAppointmentTime
      availableSlots={slots}
    />
  )
  expect(getAllByTestId('unselectedTimeButton').length).toBe(3)
})

test('<AppointmentTimes /> 1 button selected', () => {
  const { getAllByTestId } = render(
    <AppointmentTimes
      handleChange
      selectedConsultantType={'gp'}
      selectedAppointmentTime={'2019-11-27T10:11:00.000Z'}
      availableSlots={slots}
    />
  )
  expect(getAllByTestId('selectedTimeButton').length).toBe(1)
})
