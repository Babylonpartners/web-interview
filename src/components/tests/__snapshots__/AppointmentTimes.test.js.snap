// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppointmentTimes /> consultant type NOT selected 1`] = `
<p
  data-testid="appointmentTimesLoading"
>
  Loading...
</p>
`;
